### PR TITLE
Add caching

### PR DESCRIPTION
**Pull request by avivimgn**

### CODE REVIEW

File: lib/proxy.js [hunk 1]
```diff
@@ -52,19 +54,23 @@ Proxy.prototype.listen = function(options, callback = e => {}) {
   this.keepAlive = !!options.keepAlive;
   this.httpAgent = typeof(options.httpAgent) !== "undefined" ? options.httpAgent : new http.Agent({ keepAlive: this.keepAlive });
   this.httpsAgent = typeof(options.httpsAgent) !== "undefined" ? options.httpsAgent : new https.Agent({ keepAlive: this.keepAlive });
+  if (options.cache) {
+    this.cache = {
+      redisUrl: options.cache.redisUrl || 'redis://localhost:6379',
```
Comment (odedimgn) [inline lib/proxy.js:59]:
```suggestion
      redisUrl: options.cache.redisUrl || 'redis://localhost:6379/0',
```

File: lib/proxy.js [hunk 2]
```diff
@@ -52,19 +54,23 @@ Proxy.prototype.listen = function(options, callback = e => {}) {
...
+      maxTtl: options.cache.maxTtl || 60 * 60 * 3,
```
Comment (odedimgn) [inline lib/proxy.js:60]: const the 60 X 60 X 3 on the top

File: lib/proxy.js [hunk 3]
```diff
@@ -806,34 +820,39 @@ Proxy.prototype._onHttpServerRequest = function(isSSL, clientToProxyRequest, pro
...
-    const cacheableOrNormalResult = requestFunction(requestOpts, proxyToServerRequestComplete);
-    if (self.cacheRequests) {
-      cacheableOrNormalResult.on('request', onRequest);
-      cacheableOrNormalResult.on('error', self._onError.bind(self, 'CACHE_REQUEST_ERROR', ctx));
-    } else onRequest(cacheableOrNormalResult);
+    ctx.proxyToServerRequest = proto.request(ctx.proxyToServerRequestOptions, proxyToServerRequestComplete);
+    ctx.proxyToServerRequest.on('error', self._onError.bind(self, 'PROXY_TO_SERVER_REQUEST_ERROR', ctx));
+    ctx.requestFilters.push(new ProxyFinalRequestFilter(self, ctx));
+    var prevRequestPipeElem = ctx.clientToProxyRequest;
+    ctx.requestFilters.forEach(function(filter) {
```
Comment (odedimgn) [inline lib/proxy.js:851]:
```suggestion
    ctx.requestFilters.forEach((filter) => {
```

**Review by odedimgn — CHANGES_REQUESTED** [review]

File: lib/proxy.js (hunk 3 above)
Comment (avivimgn) [reply lib/proxy.js:851]: That's not my code, I'd rather keep it as is for consistency. The changes you see are caused by the revert I've done to my previous changes.

Comment (odedimgn) [reply lib/proxy.js:851]: OK

**Pull request merged**